src/Samples/SubModelSeq/Program.fs

```diff
@@ -146,113 +165,110 @@ module App =
-        |> RoseTree.addChild (createIdentifiableCounter ()) }
+        createNewLeaf ()
+        |> List.singleton
+        |> RoseTree.create dummyRootData } 
 
-  let hasId id n = n.Data.Id = id
-
-  let addChildCounter nId =
-    { Id = nId
-      Value = Counter.init }
-    |> RoseTree.addChild
+  let hasId id t = t.Data.Id = id
 
   let swapCounters swap nId =
     nId
     |> hasId
     |> List.tryFindIndex
-    |> FuncOption.bindFunc swap
+    |> FuncOption.bind swap
     |> FuncOption.inputIfNone
-    |> RoseTree.mapChildren
+
+  let updateSubtree = function
+    | CounterMsg msg -> msg |> Counter.update |> Identifiable.map |> RoseTree.mapData
+    | AddChild -> createNewLeaf () |> List.cons |> RoseTree.mapChildren
+    | Remove cId -> cId |> hasId >> not |> List.filter |> RoseTree.mapChildren
+    | MoveUp cId -> cId |> swapCounters List.swapWithPrev |> RoseTree.mapChildren
+    | MoveDown cId -> cId |> swapCounters List.swapWithNext |> RoseTree.mapChildren
 
   let update = function
     | ToggleGlobalState -> mapSomeGlobalState not
-    | CounterMsg (nId, msg) -> msg |> Counter.update |> Identifiable.applyIfId nId |> RoseTree.mapData |> mapDummyRoot
-    | AddChild pId -> Guid.NewGuid () |> addChildCounter |> Func.applyIf (hasId pId) |> RoseTree.map |> mapDummyRoot
-    | Remove nId -> nId |> hasId >> not |> List.filter |> RoseTree.mapChildren |> RoseTree.map |> mapDummyRoot
-    | MoveUp nId -> nId |> swapCounters List.swapWithPrev |> RoseTree.map |> mapDummyRoot
-    | MoveDown nId -> nId |> swapCounters List.swapWithNext |> RoseTree.map |> mapDummyRoot
-
-  /// Returns all top-level counters.
-  let topLevelCounters m =
-    m.DummyRoot |> RoseTree.dataOfChildren
-
-  /// Returns all immediate child counters of the specified parent counter ID.
-  let childCountersOf pid m =
-    m.DummyRoot
-    |> RoseTree.flatten
-    |> List.find (fun n -> n.Data.Id = pid)
-    |> RoseTree.dataOfChildren
-
-  /// Returns the parent of the specified child counter ID.
-  let parentOf cid m =
-    m.DummyRoot
-    |> RoseTree.flatten
-    |> List.find (RoseTree.dataOfChildren >> List.map Identifiable.getId >> List.contains cid)
-
-  /// Returns the sibling counters of the specified counter ID.
-  let childrenCountersOfParentOf cid =
-    cid |> parentOf >> RoseTree.dataOfChildren
+    | SubtreeMsg msg -> msg |> RoseTree.update hasId updateSubtree |> mapDummyRoot
 
 
 module Bindings =
 
   open App
 
-  let rec counterTreeBindings () : Binding<Model * Identifiable<Counter>, Msg> list = [
-    "CounterIdText" |> Binding.oneWay(fun (_, c) -> c.Id)
-
-    "CounterValue" |> Binding.oneWay(fun (_, c) -> c.Value.Count)
-    "Increment" |> Binding.cmd(fun (_, c) -> CounterMsg (c.Id, Increment))
-    "Decrement" |> Binding.cmd(fun (_, c) -> CounterMsg (c.Id, Decrement))
+  let canMoveUp (_, (p, c)) =
+    p.Children
+    |> List.tryHead
+    |> Option.map (fun c -> c.Data.Id)
+    |> Option.filter ((<>) c.Data.Id)
+    |> Option.set (MoveUp c.Data.Id)
```

Comment:
> I find the pipeline to be a very roundabout way of saying `if X then Some A else None`.

Sure.  I pushed [this commit](https://github.com/elmish/Elmish.WPF/pull/266/commits/25776153b98cc78e7f770f9d29ed3a82bd70daac) that uses pattern matching.  Is that better?

Reply:
Great!

Reply:
> I find the name `canMoveUp` to be confusing since it returns something other than `bool`.

I know what you mean.  On the other hand, abstractions exist for their consumers, and these functions are consumed by `Binding.cmdIf` like
```F#
"MoveUp" |> Binding.cmdIf(canMoveUp ...)
"MoveDown" |> Binding.cmdIf(canMoveDown ...)
```

In that context, it both reads well and is clear that it returns an optional container (specifically `Option<>`, `VOption<>`, or `Result<,>`) containing the message to be dispatched when the button is clicked.

I have been meaning to bring this up to you.  I think these three overloads are of `Binding.cmdIf` are better than the one that separately takes `exec` and `canExec` arguments.  I think the other overload is worse because it is more likely to contain a bug.  It is too easy to `exec` and `canExec` in an inconsistent manner.  [This talk](https://www.youtube.com/watch?v=z43bmaMwagI) says it much better than I can.

Reply:
> In that context, it both reads well and is clear that it returns an optional container (specifically `Option<>`, `VOption<>`, or `Result<,>`) containing the message to be dispatched when the button is clicked.

I disagree. I appreciate code that reads like well-written prose, but in my opinion, how the name `cmdIf` (or any function) meshes with the name of its argument is less important than whether the argument name (`canMoveUp`) makes sense in itself, which I think it doesn't here. In .NET, there's a strong precedent that names starting with `can`, `is`, etc. are boolean values.

It could be solved by renaming these `cmdIf` overloads to `cmdOpt`, `cmdRes` or similar, but I think that's worse since they're semantically the same as `cmdIf`.

> I think these three overloads are of `Binding.cmdIf` are better than the one that separately takes `exec` and `canExec` arguments. I think the other overload is worse because it is more likely to contain a bug. It is too easy to implement `exec` and `canExec` in an inconsistent manner. [This talk](https://www.youtube.com/watch?v=z43bmaMwagI) says it much better than I can.

I haven't watched the talk, but I agree with you. However, I don't necessarily think we should remove the `exec`/`canExec` overloads. It's a foot-gun, but they have their uses.

Reply:
> > I think these three overloads are of `Binding.cmdIf` are better than the one that separately takes `exec` and `canExec` arguments. I think the other overload is worse because it is more likely to contain a bug. It is too easy to implement `exec` and `canExec` in an inconsistent manner. [This talk](https://www.youtube.com/watch?v=z43bmaMwagI) says it much better than I can.
> 
> I haven't watched the talk, but I agree with you. However, I don't necessarily think we should remove the `exec`/`canExec` overloads. It's a foot-gun, but foot-guns have their uses and can be used responsibly.

Yes, agree.  I don't think we should removed this overload from the API.  However, I think we should subtlety suggest that users prefer the other `cmdIf` overloads by only those overloads in the samples.

Reply:
> I appreciate code that reads like well-written prose, but in my opinion, how the name `cmdIf` (or any function) meshes with the name of its argument is (generally; there's always a balance) less important than whether the argument name (`canMoveUp`) makes sense in itself, which I think it doesn't here. In .NET, there's a strong convention that names starting with `can`, `is`, etc. are boolean values.

Although I prefer I prefer the existing names, I don't mind changing them to whatever you like best.

Reply:
Since you don't mind, I suggest changing them. E.g. `moveUpMsg` (as suggested), or `tryGetMoveUpMsg`, or other suggestions if you have any.

Reply:
`moveUpMsg` and `moveDownMsg` seems good.  I made that change.